lib/TPP/ConvertPerfToLoops.cpp

```diff
@@ -0,0 +1,107 @@
+//===- ConvertPerfToLoops.cpp ------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/Perf/PerfOps.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+using namespace mlir;
+using namespace mlir::perf;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+struct ConvertBenchToLoops : public OpRewritePattern<perf::BenchOp> {
+  using OpRewritePattern<perf::BenchOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(perf::BenchOp benchOp,
+                                PatternRewriter &rewriter) const override {
+    auto loc = benchOp.getLoc();
+    auto benchYield = benchOp.getRegion().front().getTerminator();
+
+    auto numIters = rewriter.create<arith::IndexCastOp>(
+        loc, rewriter.getIndexType(), benchOp.getNumIters());
+
+    // Create benchmark loop up to perf.bench numIters.
+    auto zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    auto one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    auto loop = rewriter.create<scf::ForOp>(loc, zero, numIters, one,
+                                            benchOp.getArgs());
+    if (benchOp.getArgs().empty()) {
+      // Erase the default loop yield, it will be inserted later.
+      rewriter.eraseOp(loop.getRegion().front().getTerminator());
+    }
+
+    // Move perf.bench region inside the loop.
+    rewriter.mergeBlocks(&benchOp.getRegion().front(), loop.getBody());
+
+    // Wrap the benchmark kernel in timer calls.
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(loop.getBody());
+    auto timer = rewriter.create<perf::StartTimerOp>(
+        loc, TimerType::get(rewriter.getContext()));
+    rewriter.setInsertionPointToEnd(loop.getBody());
+    auto delta = rewriter.create<perf::StopTimerOp>(loc, rewriter.getF64Type(),
+                                                    timer.getTimer());
+
+    // Move all perf.sink ops after the timer to prevent influencing
```

Comment:
Isn't `perf.sink` just a nop?

Reply:
Ideally, it would be a nop. However, my current approach is to lower all `perf.sink`s to function calls to prevent any further compilation stages (like LLVM backend) from removing results/instructions that we want to preserve.